File: solutions/elixir/sum-of-multiples/1/lib/sum_of_multiples.ex
defmodule SumOfMultiples do
  @moduledoc false

  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    factors
    |> Enum.map(&generate(limit, &1))
    |> List.flatten()
    |> Enum.uniq()
    |> Enum.sum()
  end

  # [3, 6, 9, ...]
  @spec generate(non_neg_integer, non_neg_integer) :: [non_neg_integer]
  defp generate(_limit_number, 0), do: 0

  defp generate(limit_number, factor) do
    factor
    |> Stream.iterate(&(&1 + factor))
    |> Enum.take_while(&(&1 < limit_number))
  end
end

File: solutions/erlang/sum-of-multiples/1/src/sum_of_multiples.erl
-module(sum_of_multiples).

-export([sum/2]).

-spec sum(list(), integer()) -> integer().
sum(Factors, Limit) ->
  do_sum(Factors, Limit - 1, 0).

do_sum(_Factors, 0, Sum) ->
  Sum;
do_sum(Factors, Limit, Sum) ->
  IsMultiple = lists:any(fun(0) -> false;
                            (I) -> Limit rem I == 0 end,
                         Factors),
  case IsMultiple of
    true  -> do_sum(Factors, Limit - 1, Sum + Limit);
    false -> do_sum(Factors, Limit - 1, Sum)
  end.